### ISSUE TITLE
Do we still need the window abstractions?

### ISSUE THREAD

**Issue by mitchmindtree** [PistonDevelopers/piston#927]
I've been wondering about this for a while now so just thought I'd start a discussion!

Specifically, I want to discuss the following crates:
- [window](https://github.com/PistonDevelopers/window/blob/master/src/lib.rs)
- [glutin_window](https://github.com/PistonDevelopers/glutin_window)
- [glfw_window](https://github.com/PistonDevelopers/glfw_window)
- [sdl2_window](https://github.com/PistonDevelopers/sdl2_window)
- [piston_window](https://github.com/PistonDevelopers/piston_window)

From what I understand, Piston's window abstraction story stems from a time (just over a year ago?) when glutin was barely usable and there was only sdl2 and glfw C bindings to choose from. This lead to a lot of different people wanting to use different window libs while still wanting access to other piston libs (graphics, etc), thus `window` (along with glfw_window, sdl2_window and later glutin_window) was born.

Since then, glutin has matured a lot, builds to all major platforms (mobile is getting there), has the support of servo and from my understanding is becoming a pretty clear choice as the go-to purely rust window crate. Obviously it is nowhere near being fully stable, but it seems clear that it will continue to improve and afaik it remains the only pure rust option.

My thought process is this: Piston has spent a fairly significant amount of time maintaining its window abstractions and will probably continue to do so - perhaps we'd be better off just focusing on glutin, and use all the time saved on the rest of the Piston ecosystem and contributing to getting glutin into better shape in whatever areas it may currently lack? 

Are people still using SDL2/GLFW and if so, what are the specific reasons you use those over glutin?

**Comment by bvssvni** [comment]
There are more reasons Piston uses back-end agnostic design for the window than a pure-Rust alternative to SDL2/GLFW. See http://blog.piston.rs/2015/06/21/back-end-agnostic-design/. I would design Piston the same way if I started over today. This might be a surprise to some people! The work put into the design is quite substantial and attempted through careful thinking, and therefore should not depend that much on when it was made.

One could think of software architecture as a belief system about reality, and how well it performs depends on how it interacts with reality, but it is also a part of reality (I am reading Ben Goertzel currently so I can't help being inspired). I have tried to avoid cognitive biases, but the way human brains work I can not prevent these biases from being active and this makes it hard to present arguments that truly reflects the reality Piston has. This reality must be considered unknown and uncertain, but I consider the current architecture a better approximation than what I can come up during a few hours. The idea is that by thinking carefully and improving over time, a software architecture can become better than a single person has cognitive capacity to engage. One could point out that this might discourage discussion about software architecture, but this does not change the reality that Piston interacts with or is part of. Therefore, consider my position that there might be software that are smarter designed than people can understand on a "gut level" and rational attempts to build smarter software inevitably leads to this situation. I don't know if Piston is a such system, but there has been several cases where smart people point out things and turn out to be wrong. An argument is that do you want to keep a design that is too hard to discuss? Yet, if you rationally seek to build the smartest system possible you will end up with the same problem! :)

The arguments I want people to consider carefully, before making up their minds:

1. A cross platform library is not the same as back-end agnostic design. Back-end agnostic design is what I consider important for maintenance, and a cross platform library won't get us there. It is easier to write a window back-end than to fix the design in a cross platform library.
2. The underlying hardware for computing is going to change in the future. VR, holosystems and things we have not considered because it is not invented yet. Piston is very future oriented by design.
3. The design in Glutin does not allow custom events, which needs to propagate in a widget hierarchy. This problem is solved with Piston's event model.
4. The major part of application code is in other libraries and therefore most decisions should be considered from the impacts on the application code. A window library does not live in a vacuum. Feels hard to maintain? Software is hard and costly in general! We want to make it easier but that also depends on what you are trying to build...
5. Most positively arguments in favor of cross platform design counts as argument for back-end agnostic design, since back-end agnostic design allows both cross and native platform. The positive arguments about native platforms does not come in favor of cross platform design, but supports back-end agnostic design. Therefore, one can consider back-end agnostic design as strictly superior with a few tradeoffs. What you decide to do depends on how you weight the tradeoffs, while you can't change how cross platform and back-end design relates to each other. This part of the argument is fixed, so all counter arguments to this are logical flaws. However, this is completely unrelated to motivations for choosing a particular native or cross platform. While a design can be strictly superior to another, it does not guarantee that it will save you time for any arbitrary goal. Formally, a cross platform design solves a problem of set A, while back-end agnostic design solves a set B, where B is a fuzzy super-set of A. This is important to point out because it is here where people do the most common mistakes.

These arguments address are closely related to logical flaws in thinking, called cognitive biases, which are mistakes people do. Human thinking on a gut level is like this: You have two groups of people A and B that believes contradictory things, and you want to be friends with both. Which group will you join? You can't join both if unless you allow yourself to believe contradictory things at the same time, so you have to choose. However, this is not how the reality of software works. The human brain is not adapted to the idea that in a situation with two distinct objects, one can be a super-set of the other. 

Back-end agnostic design is a super-set of cross platform design, so when we try to make rational decisions in a such situation, the brain is simply not reliable. It is like discussing the Monty Hall problem without calculating the probability. People see Piston and they see Glutin and instinctively think of them as separate objects, but then you have already started drawing wrong conclusions! It does not help being smart, people do this mistake anyway because it is how the brain works! Even smart people do wrong things all the time!

Not only this, but if you start to see the world in a specific way, it is really hard to change you mind. So, one comes with arguments that look convincing in the hope of save the day. A typical vague argument that there is spent lot of time maintaining separate window back-ends. However, the problem with that argument is that is simply not true. The amount of time spent maintaining the window back-ends is negligible compared to the time saved when swapping out a window back-end if there is a problem. It is hard to measure this so it is not convincing if I say a factor between 10 or 100, but a vague argument does not need more than a vague counter-argument. Vague is OK. I agree that there is a possibility that a lot of time spent on maintaining separate back-ends. As a person doing this work I don't think that is the problem.

The reason I am advocating back-end agnostic design is partly because I have been bitten from this before. 10 years of work with large profit potential lost is a bitter lesson. You don't have to feel sorry for me because life turned out pretty alright after all. Yet, smart people do mistakes, and luckily I have an opportunity to not repeat the same mistake again. I think of you as a very smart person, but it would take a significant amount of evidence to change my mind, the sort of evidence that I have for my current position. I won't change it by simply by discussion. However, if you can present the best possible argument and evidence you have, then I want to consider changing my mind. There is a possibility that I am wrong and I should spend some days thinking about it.

If you don't trust me, then that is OK. Look at what other people are doing, for example Eskil Steenberg's Betray https://www.youtube.com/watch?v=oMJP6vlsmbE. It has a very interesting space of possibilities, which will be possible with the current Piston design if people want to go in this direction. I seem to remember that Eskil argued that software architecture is limiting hardware development. The argument is that new hardware is not developed because there is no market, and the market does not adapt to new hardware because the software is designed only to work on the current hardware. I think that is a reasonable argument, and it is a concern also in cross platform libraries. Back-end agnostic design solves this problem, but cross platform alone does not.

I don't think the problems I am concerned about is that important for single projects. Some people write software meant for 2-3 years for an existing market. After 3 years, there is a lot of ways you can solve such problems. However, Piston is different. Doing a such choice for the Piston project as whole means you have to think about how new hardware will work with existing libraries and whether whole ecosystems of libraries will drop out or continue working. The kind of software I am interested in building is of a different category (with whole sets of new inventions just to be able to build it). There is just a single design decision to make with a predictable cost to fix it, which is how events are handled (through pistoncore-event). UI is a big part of it and Conrod might play an important role. It is in my self interest to keep Conrod use Piston's design. I would like to see a convincing argument to switch to Glutin.

There are alternatives that could address some of the problems related to integrating with Glutin. For example, the idea of pistoncore-input was to develop a library that Glutin used such that there was a seamless integration, but @tomaka is no longer interested in this option. It might be that he have not had time to think this through, or that he wants to keep Glutin separate for other reasons. I can understand that and I am OK with it. As long as I am able to use back-end agnostic design I am happy. I am a supporter of using Glutin, but think of back-end agnostic design as important for separate reasons. He is a very smart person too! Yet, we still keep doing those silly mistakes...

tl;dr: Do I want to depend on Glutin? NO. I want a safer (and perhaps smarter?) route.

**Comment by bvssvni** [comment]
Ops. Hit the wrong button.

**Comment by bvssvni** [comment]
The major cost is that generic libraries need more constraints. This can be fixed in application code using the backend library https://github.com/pistondevelopers/backend if there are many relationships between the types and they form a closed type family. This library is still experimental. For generic libraries there is no similar solution, so the code gets more verbose.

Library dependencies follow the same laws as genes in biology, where a gene must be universal before other genes can build further complexity from it. This is why evolution is slow and makes incremental progress in fitness. This is not true for software in the sense that you can build an abstraction that hides the complexity and use modular design. However, it is true in the sense that people share code, and they have to rewrite parts of it to make it work. Some aspects of this in cross platform libraries makes it harder to share modified versions. I might be wrong about the scope of this problem for window abstractions.

It is very likely that people will use Glutin more than other back-ends. Yet, the widespread use of SDL2 in the game industry is a motivation to use back-end agnostic design. In addition we have a stronger position for taking advantage of new hardware.

In the future, we might use interactive applications contained within a bigger application. This requires a common event model. You often start building a reusable component in one project, and then figure out this solves a problem in another project. At the bottom level you have libraries for this pattern, but at higher level the components spans over multiple libraries. Without an abstraction, it is frequently easier to write things from scratch than reusing code. If people are used to writing things from scratch, they feel like they make a lot of progress, but in the end designing for reusing code wins. At product level you have a whole package of complexity that you want to move from A to B, and this becomes far easier if you can switch out the back-end.

Yet, I am a bit surprised of how often people questioning the window abstraction. From one point of view I agree with lot of the arguments, but there are concerns that are not that easy to see at first. I don't spend much time thinking about those concerns, because the window abstraction solves it, but each time the question comes up I need to think through it again to explain it. I don't know how to measure or compare this objectively because I don't write that much platform specific code. However, it doesn't look like the concerns I have gets addressed in a satisfying way.

The thing that will most likely change my mind is something I have not yet thought about. It could be that generic libraries become too hard to maintain, but I don't find that argument convincing now. I don't expect the maintenance of the window back-ends in particular to be the problem, but expect this to be something to think about for generic libraries in general.

My position is still that back-end agnostic design is the direction to go.

**Comment by bvssvni** [comment]
I just had an idea https://github.com/PistonDevelopers/piston/issues/933 of how to decouple the event model and the window abstraction, which might simplify the maintenance of generic libraries.

**Comment by bvssvni** [comment]
I'll implemented the idea and then we'll see how it goes.

Closing.
